### DISCUSSION TITLE
EVP_DigestSignInit tries to initialize engine even though it was specified as null

### DISCUSSION THREAD

**Discussion by njbelf** [openssl/openssl#21932] [Q&A] [answered]
I have successfully compiled OpenSSL with the enable_fips flag  on Windows with the WIN32A target.  I should note that WIN64A fails to build with errors in the makefile which I haven't had time to go figure out.

The resultant openssl.exe operates fully and passes all tests from the nmake tests command.

I have an application which provides signing services

I am attempting to utilize EVP_DigestSignInit with the following call:

```
if (EVP_DigestSignInit(mdcxt, NULL, EVP_sha3_384(), NULL, privateKey) != 1)
{
    // handle error
}
```
This generates the following exception:

Exception thrown: read access violation.
**e** was 0x88004700.
 
and here's the stack:

	libcrypto-3.dll!ENGINE_init(engine_st * e) Line 94	C
 	libcrypto-3.dll!int_ctx_new(ossl_lib_ctx_st * libctx, evp_pkey_st * pkey, engine_st * e, const char * keytype, const char * propquery, int id) Line 240	C
 	libcrypto-3.dll!EVP_PKEY_CTX_new_from_pkey(ossl_lib_ctx_st * libctx, evp_pkey_st * pkey, const char * propquery) Line 374	C
 	libcrypto-3.dll!do_sigver_init(evp_md_ctx_st * ctx, evp_pkey_ctx_st * * pctx, const evp_md_st * type, const char * mdname, ossl_lib_ctx_st * libctx, const char * props, engine_st * e, evp_pkey_st * pkey, int ver, const ossl_param_st * params) Line 67	C
 	libcrypto-3.dll!EVP_DigestSignInit(evp_md_ctx_st * ctx, evp_pkey_ctx_st * * pctx, const evp_md_st * type, engine_st * e, evp_pkey_st * pkey) Line 386	C
 	Encrypt&SignServer.exe!ESSCommandsClass::Sign(std::basic_string<unsigned char,std::char_traits<unsigned char>,std::allocator<unsigned char>> x, unsigned int dataLen, unsigned short sessionID) Line 335	C++

My question is why is **e** anything but NULL?

**Comment by mattcaswell** [comment]
Which version of OpenSSL are you using?

**Comment by mattcaswell** [comment]
How was privateKey created? The value of `e` will either come from an explicit ENGINE that you pass in to the `EVP_DigestSignInit` call *or* it will use an ENGINE that was already associated with the pkey if there is one.

**Comment by t8m** [comment]
Or a default engine can be used if it is configured by the OpenSSL config file.

**Comment by njbelf** [comment]
I'm using OpenSSL 3.0.10, sorry I had intended to write that originally.   

The private key was fetched from a PEM file through BIO_new_file(<path>, "r"); and PEM_read_bio_PrivateKey.

There is no engine associated with this key as it's a plain private key.

I've searched high and low to figure out how to use an engine in this context and I'm none the wiser.   Could you give an example of default engine being configured in the config file?

**Comment by mattcaswell** [comment]
> Could you give an example of default engine being configured in the config file?

See this page, and in particular the "Engine Configuration" section
https://www.openssl.org/docs/man3.0/man5/config.html

You might want to examine the state of `privateKey` in a debugger and determine whether `privateKey->engine` or `privateKey->pmeth_engine` are set.

**Comment by njbelf** [comment]
Ah, therein lies a rub.    An EVP_PKEY* cannot be examined in a debugger because evp_pkey_st is an opaque type.    I don't think privateKey->engine or privateKey->pmeth_engine should be set since this is a plain private key.    

However, it's obvious that those values _are_ being set with random data.

**Comment by njbelf** [comment]
and there is no default engine in the Openssl installed directory.   My engines-3 directory only contains capi.dll, loader_attic.dll and padlock.dll

**Comment by njbelf** [comment]
Here's what *pkey looks like in do_sigver_init:

![pkeyvalues](https://github.com/openssl/openssl/assets/12550054/30918828-be4c-4097-ad1a-17aae138304b)

**Comment by njbelf** [comment]
I don't see any data from the PEM file in that pkey

**Comment by mattcaswell** [comment]
On first glance that key looks completely screwed. Could it be corrupted? Have you run valgrind/asan etc?

**Comment by mattcaswell** [comment]
In fact it has a ref count of 0 which is invalid (possibly this is a use-after-free?)

**Comment by njbelf** [comment]
We were able to use this key when we were using 1.0.1m.

**Comment by njbelf** [comment]
I think the code:
```
key_bio = BIO_new_file(<path>, "r");
private_key = PEM_read_bio_PrivateKey(key_bio, &private_key, 0, NULL);
```
Doesn't properly read a PEM private key.

**Comment by mattcaswell** [comment]
> We were able to use this key when we were using 1.0.1m.

It's almost certainly not the underlying key data that is corrupt, but the in memory representation of it (so the fact that it worked in 1.0.1m is not the point). It is very likely that something in your application is causing it to become corrupted in some way.

> Doesn't properly read a PEM private key.

It might be useful to look at the key in the debugger immediately before `PEM_read_bio_PrivateKey` returns. You should at least see the `references` field be a value that is >= 1.

**Reply by njbelf** [reply, marked as answer]
I found that there was code that was taking the address of the address of the EVP_PKEY structure and that was causing the corrupted key you saw above.   I have corrected that code and lo, and behold, the code works as expected.     The key was checking that references field.  I had to get the memory viewer going to see the data and that showed the wrong address being given.    Thank you for the education on how this stuff really works.

**Reply by mattcaswell** [reply]
Great. Glad you fixed it.
